### PR TITLE
Add `recurse` method to `TailCalls`

### PR DESCRIPTION
**Pull request by tmccombs**
This method is a simple convenience that allows you to recursively call an anonymous function, by passing a wrapper around that function as an argument. This can be used as a looping construct.

### CODE REVIEW

File: src/library/scala/util/control/TailCalls.scala [hunk 1]
```diff
@@ -111,4 +111,31 @@ object TailCalls {
...
+   * This can be used to implement a loop similar to using self-recursion
+   * with an anonymous function.
+   *
+   * Fore example computing a fibonacci number could be implemented as:
+   *
+   * {{{
+   * import scala.util.control.TailCalls._
+   *
+   * def fib(n: Int): Int =
+   *    recurse[(Int, Int, Int), Int]((n, 1, 0)) ( recur => { case (i, n1, n2) =>
```
Comment (tmccombs) [inline src/library/scala/util/control/TailCalls.scala]: It's unfortunate that you have to explicitly give the type parameters, but I wasn't able to figure out a way to get scala to infer the types (at least in scala 2, maybe 3 can do it?)

Needing an extra set of braces around the case is a little annoying too.

Comment (bishabosha) [reply src/library/scala/util/control/TailCalls.scala]: You can make a Recursor class that captures the initial state and infers the type, then that class has an apply method accepts the lambda

**Comment by SethTisue** [comment]
This looks plausible, but we cannot currently add new methods to the Scala 2 standard library, as per https://github.com/scala/scala-dev/issues/661

I've added the "library:next" label so we could revisit this once the situation changes.

**Comment by tmccombs** [comment]
FWIW, in scala 3 it might be worth implementing with a macro, so it can take an arbitrary number of arguments instead of exactly 1

**Pull request closed without merging**